src/Hosting/Systemd/src/SystemdHelpers.cs

```diff
@@ -12,14 +16,55 @@ public static class SystemdHelpers
     {
         private const string INVOCATION_ID = "INVOCATION_ID";
 
+        private static bool? _isSystemdService;
+
         /// <summary>
         /// Check if the current process is hosted as a systemd Service.
         /// </summary>
         /// <returns><c>True</c> if the current process is hosted as a systemd Service, otherwise <c>false</c>.</returns>
         public static bool IsSystemdService()
+            => _isSystemdService ?? (bool)(_isSystemdService = CheckSystemdUnit());
+
+        private static bool CheckSystemdUnit()
         {
-            // We use the INVOCATION_ID envirionment variable that was introduced in systemd 232 (released 2016-11-03).
-            return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(INVOCATION_ID));
+            // No point in testing anything unless it's Unix
+            if (Environment.OSVersion.Platform != PlatformID.Unix)
+            {
+                return false;
+            }
+
+            // We've got invocation id, it's systemd >= 232 running a unit
+            if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable(INVOCATION_ID)))
```

Comment:
note: for newer versions of systemd, child processes will also return true because INVOCATION_ID is set.
For older version of systemd, only the direct child returns true.

Reply:
Pushed potentially improved comments.